internal/pkg/gateway/registry.go

```diff
@@ -93,29 +93,51 @@ func (reg *registry) endorsers(channel string, chaincode string) ([]*endorser, e
 				}
 				endpoint := msg.GetAliveMsg().Membership.Endpoint
 
-				groupPeers = append(groupPeers, &endorserState{peer: peer, endpoint: endpoint, height: height})
+				// find the endorser in the registry for this endpoint
+				var endorser *endorser
+				if endpoint == reg.localEndorser.address {
+					endorser = reg.localEndorser
+				} else if e, ok := reg.remoteEndorsers[endpoint]; ok {
+					endorser = e
+				} else {
+					reg.logger.Warnf("Failed to find endorser at %s", endpoint)
+					continue
+				}
+
+				// add to list of endorsing peers, unless endorsingOrgs is specified and it's not in one of the allowed orgs
+				if len(endorsingOrgs) == 0 || contains(endorsingOrgs, endorser.mspid) {
+					groupPeers = append(groupPeers, &endorserState{peer: peer, endorser: endorser, endpoint: endpoint, height: height})
```

Comment:
The endorsers service discovery query takes into consideration both the chaincode level and collection level endorsement policy. But Manish is correct, we need to discuss the semantics of endorsingOrgs.

I think the primary use case is when the application knows exactly which orgs to target. In that case, the Service Discovery peers query should be utilized, rather than the more expensive endorsers query. This will often be the case when using state-based endorsement (as SD endorsers query doesn't consider state-based endorsement requirements). See more history in https://jira.hyperledger.org/browse/FABN-1601.

Another use case is when the endorsement policy should be used (for example 3 out of 100 orgs), but you want to target some specific orgs, for example you want to utilize OrgA and OrgB but let service discovery pick any third org.

Bret implemented both approaches for the node SDK, and I think we have to do the same for Gateway.

A note on the endorsers query org filtering approach... Bret used a client side endorsers query filter in node SDK , but that approach doesn't work at all when there are many layouts returned since service discovery may only return a random subset. That means you may entirely miss on the intersection of layouts returned and orgs required. The filter therefore must be done on the service discovery side - we need to consider extending the SD endorsers query to take the list of required orgs.  In fact, this is why we have a single team responsible for Gateway and Service Discovery now.

Reply:
I've reimplemented this as Dave suggests to return all the peers from the list of 'endorsingOrganizations' that have the given chaincode installed.